File: android5/app/src/main/java/utils/Tasker.kt
/*
 * This file is part of Blokada.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Copyright © 2022 Blocka AB. All rights reserved.
 *
 * @author Karol Gusak (karol@blocka.net)
 */

package utils

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.*
import model.BlokadaException
import repository.Repos

val DEFAULT_USER_INTERACTION_DEBOUNCE = 300L

internal data class TaskResult<T, Y> (
   val argument: T,
   val result: Y?,
   val error: BlokadaException?
)

/**
 * This class is an almost-direct port of the same class from our iOS code. This is far from ideal
 * and I'm sure it can be done better with Kotlin coroutines.
 */
open class Tasker<T, Y>(
    internal val owner: String,
    private val debounce: Long = DEFAULT_USER_INTERACTION_DEBOUNCE,
    private val errorIsMajor: Boolean = false
) {

    internal val processingRepo by lazy { Repos.processing }

    internal val writeRequests = MutableSharedFlow<T?>()
    private val requests = writeRequests.filterNotNull()

    private val writeResponses = MutableSharedFlow<TaskResult<T, Y>?>()
    internal val responses = writeResponses.filterNotNull()

    // This will block coroutine for lifetime of this task handler.
    suspend fun setTask(task: suspend (T) -> Y) {
        var publisher = requests
        if (debounce > 0) {
            publisher = requests.debounce(debounce)
        }
        publisher.collect {
            try {
                Logger.v("Tasker", "$owner: $it")
                val result = task(it)
                writeResponses.emit(TaskResult(it, result, null))
                processingRepo.notify(owner, ongoing = false)
            } catch (ex: BlokadaException) {
                writeResponses.emit(TaskResult(it, null, ex))
                processingRepo.notify(owner, ex, major = errorIsMajor)
            } catch (err: Throwable) {
                val ex = BlokadaException("task error: $owner", err)
                writeResponses.emit(TaskResult(it, null, ex))
                processingRepo.notify(owner, ex, major = errorIsMajor)
            }
        }
    }

    suspend fun send(argument: T): Y {
        processingRepo.notify(owner, ongoing = true)
        val resp = GlobalScope.async {
            // Just return first result if debounce is set, as it's used for cases when we
            // only care about the latest invocation in a short time.
            responses.first { debounce != 0L || argument == it.argument }
        }
        writeRequests.emit(argument)
        val r = resp.await()

        val error = r.error
        if (error != null) {
            throw error
        }

        return r.result as Y
    }

}

class SimpleTasker<Y>(
    owner: String,
    debounce: Long = DEFAULT_USER_INTERACTION_DEBOUNCE,
    errorIsMajor: Boolean = false
) : Tasker<Ignored, Y>(owner, debounce, errorIsMajor) {

    suspend fun send(): Y {
        processingRepo.notify(owner, ongoing = true)
        val resp = GlobalScope.async {
            responses.first()
        }
        writeRequests.emit(true)
        val r = resp.await()

        val error = r.error
        if (error != null) {
            throw error
        }

        return r.result as Y
    }

}

File: ios/App/Util/Tasker.swift
//
//  This file is part of Blokada.
//
//  This Source Code Form is subject to the terms of the Mozilla Public
//  License, v. 2.0. If a copy of the MPL was not distributed with this
//  file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
//  Copyright © 2021 Blocka AB. All rights reserved.
//
//  @author Karol Gusak
//

import Foundation
import Combine

fileprivate struct TaskResult<T: Equatable, Y> {
    let argument: T
    let result: Y?
    let error: Error?
}

class Tasker<T: Equatable, Y> {

    fileprivate lazy var requests = PassthroughSubject<T, Never>()
    fileprivate lazy var response = PassthroughSubject<TaskResult<T, Y>, Never>()

    fileprivate lazy var processingRepo = Repos.processingRepo

    fileprivate let owner: String
    private let debounce: Double
    private let errorIsMajor: Bool
    private let bgQueue = DispatchQueue(label: "TaskerBgQueue")
    private var cancellable: AnyCancellable? = nil

    init(_ owner: String = "Unknown", debounce: Double = DEFAULT_USER_INTERACTION_DEBOUNCE, errorIsMajor: Bool = false) {
        self.debounce = debounce
        self.owner = owner
        self.errorIsMajor = errorIsMajor
    }

    func setTask(_ task: @escaping (T) -> AnyPublisher<Y, Error>) {
        var publisher: AnyPublisher<T, Never> = requests.eraseToAnyPublisher()
        if debounce > 0 {
            publisher = requests
            .debounce(for: .seconds(debounce), scheduler: bgQueue)
            .eraseToAnyPublisher()
        }

        cancellable = publisher
        .receive(on: bgQueue)
        .map { argument in
            BlockaLogger.v("Tasker", "\(self.owner): \(argument)")
            return argument
        }
        .flatMap { argument in
            task(argument)
            .tryMap { result in
                TaskResult(argument: argument, result: result, error: nil)
            }
            .catch { err in
                return Just(TaskResult(argument: argument, result: nil, error: err))
            }
        }
        .sink(
            onValue: { it in
                if let err = it.error {
                    // TODO: what about major errors
                    self.processingRepo.notify(self.owner, err, major: self.errorIsMajor)
                } else {
                    self.processingRepo.notify(self.owner, ongoing: false)
                }
                self.response.send(it)
            }
        )
    }

    func send(_ argument: T) -> AnyPublisher<Y, Error> {
        processingRepo.notify(owner, ongoing: true)
        let responsePub = response
        // Just return first result if debounce is set, as it's used for cases when we
        // only care about the latest invocation in a short time.
        .first { it in self.debounce != 0.0 || it.argument == argument }
        .tryMap { it -> Y in
            if let err = it.error {
                throw err
            }
            return it.result!
        }
        .eraseToAnyPublisher()
        requests.send(argument)
        return responsePub
    }

}

class SimpleTasker<Y>: Tasker<Bool, Y> {

    override init(_ owner: String = "Unknown", debounce: Double = DEFAULT_USER_INTERACTION_DEBOUNCE, errorIsMajor: Bool = false) {
        super.init(owner, debounce: debounce, errorIsMajor: errorIsMajor)
    }

    func send() -> AnyPublisher<Y, Error> {
        processingRepo.notify(owner, ongoing: true)
        let responsePub = response
        .first()
        .tryMap { it -> Y in
            if let err = it.error {
                throw err
            }
            return it.result!
        }
        .eraseToAnyPublisher()
        requests.send(true)
        return responsePub
    }

}
